ehrql/permissions.py

```diff
@@ -0,0 +1,145 @@
+    pass
+
+
+def claim_permissions(*permissions):
+    """
+    This function allows you to access any restricted table or feature when working with
+    dummy data. It will NOT allow you access with real data: for that you will need the
+    appropriate permissions on the OpenSAFELY platform.
+
+    Permission names are strings and should be written with double quotes e.g.
+
+        from ehrql import claim_permissions
+
+        claim_permissions("some_permission", "another_permission")
+
+    This can go anywhere in your dataset or measure definition file.
+
+    You can make multiple `claim_permissions()` calls and the permissions will be
+    combined together.
+    """
+    for permission in permissions:
+        # Use dictionary as convenient ordered set
+        CLAIMED_PERMISSIONS[permission] = True
+
+
+def clear_claimed_permissions():
+    CLAIMED_PERMISSIONS.clear()
+
+
+def get_claimed_permissions():
+    return tuple(CLAIMED_PERMISSIONS.keys())
+
+
+def enforce_permissions(dataset, environ):
+    available_permissions = parse_permissions(environ)
+    missing_permissions = get_missing_permissions(dataset, available_permissions)
+    if missing_permissions:
+        raise EHRQLPermissionError(
+            f"You do not currently have all the permissions needed for this action.\n"
+            f"\n"
+            f"Missing permissions are:\n"
+            f"\n"
+            f"{format_permission_list(missing_permissions)}\n"
+            f"\n"
+            f"If you think this is a mistake and that you should have these "
+            f"permissions please contact OpenSAFELY support."
+        )
+
+
+def enforce_permissions_for_dummy_data(dataset, claimed_permissions):
+    missing_permissions = get_missing_permissions(dataset, claimed_permissions)
+    if missing_permissions:
+        claim_list = ", ".join(
+            f'"{p}"' for p in [*claimed_permissions, *missing_permissions]
+        )
+        message = (
+            f"Some of the tables or features you are using require special permission to use with real\n"
+            f"patient data. The permissions needed are:\n"
+            f"\n"
+            f"{format_permission_list(missing_permissions)}\n"
+            f"\n"
+            f"You can continue to work on your code using dummy data by “claiming” "
+            f"the required permisions:\n"
+            f"\n"
+            f"    from ehrql import claim_permissions\n"
+            f"    claim_permissions({claim_list})\n"
+            f"\n"
+            f"Note that you will only be able to run your code against real data if you actually have these\n"
+            f"permissions assigned by the OpenSAFELY team. For more information see:\n"
+            f"https://docs.opensafely.org/ehrql/reference/language/#permissions"
+        )
+        # For the inital rollout of this feature we issue a warning locally but continue
+        # running. Eventually we want to make this a hard error so that it can't be
```

Comment:
:+1: 

Is doing this part of #2515 or should we make a new issue to make sure we don't forget it?